### PR TITLE
Add ChangeView.Reason

### PR DESCRIPTION
**Pull request by shargon**
TODO: We can review if the most common reason was `TxNotFound` and reduce the next block proposal from 500 tx to 250.

### CODE REVIEW

File: neo/Consensus/ChangeView.cs [hunk 1]
```diff
@@ -4,31 +4,49 @@ namespace Neo.Consensus
 {
     public class ChangeView : ConsensusMessage
     {
+        public enum ChangeViewReason : byte
+        {
+            Timeout = 0x0,
+            WrongExpectedView = 0x1,
+            TxNotFound = 0x2,
+            TxRejectedByPolicy = 0x3,
+            TxInvalid = 0x4,
```
Comment (vncoelho) [inline neo/Consensus/ChangeView.cs]: Veri gud veri naiceee, Shargon!

I like this. It is great that you are pushing this!

**Comment by shargon** [comment]
I think that it would be good for take decisions according to the view change. For example, if we have 5 view changes because policy, maybe i need to clean my memory pool?

**Comment by lock9** [comment]
@shargon  That is a good idea! When you had that method to clean added to the ProtocolHandler, I already had in mind that this "clean" should be automated, just like most apps are (if you get a memory warning, you need to clean resources).

So, yes, I think that this "inconsistency" of the network may lead nodes to clean their data. 
I'm just a little skeptical because @canesin  said something about "having low risk transactions that can be 'confirmed' without waiting for the block".
If we have this "cleanup", there is a chance these 'valid transactions' are dropped.

What do you guys think? I think this is a good idea, not sure about side effects, especially for Nash

File: neo/Consensus/ChangeView.cs [hunk 2]
```diff
@@ -4,31 +4,49 @@ namespace Neo.Consensus
 {
     public class ChangeView : ConsensusMessage
     {
+        public enum ChangeViewReason : byte
```
Comment (erikzhang) [inline neo/Consensus/ChangeView.cs]: It should be put into a separate file.

Comment (shargon) [reply neo/Consensus/ChangeView.cs]: Done!

File: neo/Consensus/ChangeView.cs [hunk 3]
```diff
@@ -8,28 +8,37 @@ public class ChangeView : ConsensusMessage
         /// NewViewNumber is always set to the current ViewNumber asking changeview + 1
         /// </summary>
         public byte NewViewNumber => (byte)(ViewNumber + 1);
+
+        /// <summary>
+        /// Reason
+        /// </summary>
+        public ChangeViewReason Reason;
```
Comment (erikzhang) [inline neo/Consensus/ChangeView.cs]: This order should be change too.

Comment (shargon) [reply neo/Consensus/ChangeView.cs]: done

File: neo/Consensus/ConsensusService.cs [hunk 4]
```diff
@@ -124,7 +124,7 @@ private void CheckExpectedView(byte viewNumber)
                     // Communicate the network about my agreement to move to `viewNumber`
                     // if my last change view payload, `message`, has NewViewNumber lower than current view to change
                     if (message is null || message.NewViewNumber < viewNumber)
-                        localNode.Tell(new LocalNode.SendDirectly { Inventory = context.MakeChangeView() });
+                        localNode.Tell(new LocalNode.SendDirectly { Inventory = context.MakeChangeView(ChangeViewReason.WrongExpectedView) });
```
Comment (erikzhang) [inline neo/Consensus/ConsensusService.cs]: I think it sends `ChangeView` because everyone wants to change view. There is nothing wrong.

**Comment by vncoelho** [comment]
Hi @erikzhang Yoda master,

I think that if the reason is something like invalid tx there will be no reason for recovery, for example.

File: neo/Consensus/ConsensusService.cs (hunk 4 above)
Comment (vncoelho) [reply neo/Consensus/ConsensusService.cs]: Hi @erikzhang Yoda master,

I think that if the reason is something like invalid tx there will be no reason for recovery, for example.

Comment (shargon) [reply neo/Consensus/ConsensusService.cs]: `ConsensusChange`?

Comment (erikzhang) [reply neo/Consensus/ConsensusService.cs]: `Follow`?

Comment (vncoelho) [reply neo/Consensus/ConsensusService.cs]: @erikzhang, my comment above was wrong...

I just remembered now that we already created a special payload `RecoveryRequest`.
In this sense, I am not sure about a real use of this Reason.

Should we revert, @shargon? Or do you have an idea for using it?
@igormcoelho

**Pull request merged**